transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollSocketChannel.java

```diff
@@ -198,22 +194,46 @@ protected AbstractEpollUnsafe newUnsafe() {
         return new EpollSocketChannelUnsafe();
     }
 
+    private InetSocketAddress computeRemoteAddr(InetSocketAddress remoteAddr) {
+        InetSocketAddress osRemoteAddr = fd().remoteAddress();
+        if (osRemoteAddr != null) {
+            if (PlatformDependent.javaVersion() >= 7) {
+                try {
+                    // Only try to construct a new InetSocketAddress if we using java >= 7 as getHostString() does not
+                    // exists in earlier releases and so the retrieval of the hostname could block the EventLoop if a
+                    // reverse lookup would be needed.
+                    return new InetSocketAddress(InetAddress.getByAddress(remoteAddr.getHostString(),
+                            osRemoteAddr.getAddress().getAddress()),
+                            osRemoteAddr.getPort());
+                } catch (UnknownHostException ignore) {
+                    // Should never happen but fallback to osRemoteAddr anyway.
+                }
+            }
+            return osRemoteAddr;
+        }
+        return remoteAddr;
+    }
+
     @Override
     protected boolean doConnect(SocketAddress remoteAddress, SocketAddress localAddress) throws Exception {
         if (localAddress != null) {
             checkResolvable((InetSocketAddress) localAddress);
         }
         checkResolvable((InetSocketAddress) remoteAddress);
+        InetSocketAddress remoteAddr = (InetSocketAddress) remoteAddress;
         boolean connected = super.doConnect(remoteAddress, localAddress);
         if (connected) {
-            remote = (InetSocketAddress) remoteAddress;
+            remote = computeRemoteAddr(fd().remoteAddress());
             return true;
+        } else {
+            // Store for later usage n doFinishConnect()
+            requestedRemote = remoteAddr;
         }
         // We always need to set the localAddress even if not connected yet
```

Comment:
actually yes I think we should better do this in all cases (even if we not did before as well).